packages/langium/src/parser/langium-parser.ts

```diff
@@ -283,15 +283,20 @@ export class LangiumParser extends AbstractLangiumParser {
     private assignMandatoryProperties(obj: any): void {
         const typeMetaData = this.astReflection.getTypeMetaData(obj.$type);
         for (const mandatoryProperty of typeMetaData.mandatory) {
-            const value = obj[mandatoryProperty.name];
-            if (mandatoryProperty.type === 'array' && !Array.isArray(value)) {
-                obj[mandatoryProperty.name] = [];
-            } else if (mandatoryProperty.type === 'boolean' && value === undefined) {
-                obj[mandatoryProperty.name] = false;
+            if (obj[mandatoryProperty.name] === undefined) {
+                obj[mandatoryProperty.name] = this.copyDefaultValue(mandatoryProperty.defaultValue!);
             }
         }
     }
 
+    private copyDefaultValue(propertyType: MandatoryPropertyType): MandatoryPropertyType {
+        if (Array.isArray(propertyType)) {
+            return [...propertyType.map(this.copyDefaultValue)];
```

Comment:
The method is used with a function reference here. I suggest moving this to a plain function to avoid bugs due to unbound `this` (otherwise you'd need to write `this.copyDefaultValue.bind(this)`).